src/utils/token-verifier.ts

```diff
@@ -0,0 +1,290 @@
+  errorType: new (info: ErrorInfo, message?: string) => PrefixedFirebaseError;
+}
+
+/**
+ * Class for verifying general purpose Firebase JWTs. This verifies ID tokens and session cookies.
+ */
+export class FirebaseTokenVerifier {
+  private publicKeys: {[key: string]: string};
+  private publicKeysExpireAt: number;
+  private readonly shortNameArticle: string;
+
+  constructor(private clientCertUrl: string, private algorithm: jwt.Algorithm,
+              private issuer: string, private tokenInfo: FirebaseTokenInfo,
+              private readonly app: FirebaseApp) {
+
+    if (!validator.isURL(clientCertUrl)) {
+      throw new Error('The provided public client certificate URL is an invalid URL.');
+    } else if (!validator.isNonEmptyString(algorithm)) {
+      throw new Error('The provided JWT algorithm is an empty string.');
+    } else if (!validator.isURL(issuer)) {
+      throw new Error('The provided JWT issuer is an invalid URL.');
+    } else if (!validator.isNonNullObject(tokenInfo)) {
+      throw new Error('The provided JWT information is not an object or null.');
+    } else if (!validator.isURL(tokenInfo.url)) {
+      throw new Error('The provided JWT verification documentation URL is invalid.');
+    } else if (!validator.isNonEmptyString(tokenInfo.verifyApiName)) {
+      throw new Error('The JWT verify API name must be a non-empty string.');
+    } else if (!validator.isNonEmptyString(tokenInfo.jwtName)) {
+      throw new Error('The JWT public full name must be a non-empty string.');
+    } else if (!validator.isNonEmptyString(tokenInfo.shortName)) {
+      throw new Error('The JWT public short name must be a non-empty string.');
+    } else if (!(typeof tokenInfo.errorType === 'function' && tokenInfo.errorType !== null)) {
+      throw new Error('The provided error type must be a non-null PrefixedFirebaseError type.');
+    } else if (!validator.isNonNullObject(tokenInfo.errorCodeConfig) ||
+      !('invalidArg' in tokenInfo.errorCodeConfig ||
+        'invalidCredential' in tokenInfo.errorCodeConfig ||
+        'internalError' in tokenInfo.errorCodeConfig ||
+        'expiredError' in tokenInfo.errorCodeConfig)) {
+      throw new Error('The provided error code config must be a non-null ErrorCodeInfo object.');
+    }
+    this.shortNameArticle = tokenInfo.shortName.charAt(0).match(/[aeiou]/i) ? 'an' : 'a';
+
+    // For backward compatibility, the project ID is validated in the verification call.
+  }
+
+  /**
+   * Verifies the format and signature of a Firebase JWT token.
+   *
+   * @template T The returned type of the decoded token.
+   * @param {string} jwtToken The Firebase JWT token to verify.
+   * @param {boolean=} isEmulator Whether to accept Emulator tokens.
+   * @return {Promise<T>} A promise fulfilled with the decoded claims of the Firebase Auth ID
+   *                           token.
+   */
+  public verifyJWT<T extends object>(jwtToken: string, isEmulator = false): Promise<T> {
```

Comment:
To tighten the type check we can introduce a base type for returned tokens...
```
interface FirebaseToken {
  // registered JWT claims
  iss: string
  aud: string
}

interface DecodedIdToken extends FirebaseToken {
...
}

public verifyJWT<T extends FirebaseToken>
```

Let me know your thoughts. Thanks!

Reply:
Yeah, this looks good to me. But I don't think we need a generic at all. Just make auth extend the TokenVerifier class, and transform the FirebaseToken into DecodedIdToken there.